README.md
# easy_anim

💨🚀 用一种简单的方式，声明式描述使用Flutter的动画 


## 前言

> 谈到Flutter动画的时候，我们一般想到的是什么？
> 
> 如果你照着Flutter文档手册看完，脑子里肯定一堆Tween, AnimationController, forward()等等乱七八糟的东西。写肯定是能写，但是非常影响开发效率，并且增加代码的复杂度，如果维护不好很容易就会让代码乱掉
> 
> 当然，官方现在也有提供隐式动画如TweenAnimationBuilder可以实现一些简单的单个动画，还有TweenSequence实现串行动画，但是我实际项目中最简单的动画也是缩放同时旋转的并行动画（又称交织动画），所以并不适用，在pub.dev、Github找了几个小时第三方库也没有找到可以声明式的方式简单的方式写动画的库，所以决定自己撸一个
> 


## 优点
- 使用组件方式，声明式描述动画
- 自维护controller，使用动画代码简洁
- 支持延时执行动画
- 支持循环执行动画
- 支持并行动画、串行动画以及串行并行动画




## 快速开始

### 添加依赖
``` yaml
dependencies:
  easy_anim: ^2.1.1
```


### 一个极简使用示例
``` dart
import 'package:easy_anim/easy_anim.dart';

EasyTweenAnimation(
  animSequence: [
    EasyTweenAnimationItem(
        animatables: {
          "width": Tween(begin: 0.0, end: 200.0),
        },
        weight: 100,
    ),
  ],
  duration: Duration(seconds: 2),
  builder: (BuildContext context, CurvedAnimation curvedAnimation, Map<String, Animation> animationMap, AnimationController animationController, Widget child){

    Animation width = animationMap['width'];

    return AnimatedBuilder(animation: curvedAnimation, builder: (context, child){
          return Container(
            width: width.value,
            height: 100,
            decoration: BoxDecoration(
              color: Colors.red
            ),
          );
        },
    );
  }
)
```


### example工程截图
查看截图效果的对应源码：[example/lib/main.dart](example/lib/main.dart)

<img src="doc/assets/demo1.gif" width="300" > <img src="doc/assets/demo2.gif" width="300" >



### 使用示例
``` dart
 import 'package:easy_anim/easy_anim.dart';
 
 EasyTweenAnimation(
   animSequence: [
     EasyTweenAnimationItem(
      animatables: {
        "angle": Tween<double>(begin: 0, end: 1 * pi),
        "color": ColorTween(begin: Colors.blue, end: Colors.blue),
        "width": Tween<double>(begin: 100, end: 100),
      },
      weight: 50.0, // 占总时长的50%，也就是2秒的0%~50%时的动画效果
    ),


    EasyTweenAnimationItem(
      animatables: {
        "color": ColorTween(begin: Colors.blue, end: Colors.red),
        "width": Tween<double>(begin: 100, end: 200),
      },
      weight: 50.0, // 占总时长的50%，也就是2秒的50%~100%时的动画效果
    ),
  ],

  duration: Duration(seconds: 2), // 执行动画总时长
  delay: Duration(milliseconds: 200), // 延时多久执行，默认0秒立即执行
  loop: true, // 是否循环执行动画
  builder: (BuildContext context, CurvedAnimation curvedAnimation, Map<String, Animation> animationMap, AnimationController animationController, Widget child){

    // 从animationMap里取出各个效果的Animation对象
    Animation angle = animationMap["angle"];
    Animation color = animationMap['color'];
    Animation width = animationMap['width'];

    // AnimatedBuilder 里使用 各个效果的Animation对象value即可
    return AnimatedBuilder(animation: curvedAnimation, builder: (context, child){
      return Transform.rotate(
        angle: angle.value,
        child: Container(
          color: color.value,
          width: width.value,
          height: width.value,
        ),
      );
    });
  },
)

```

### 参数说明

#### EasyTweenAnimation 组件参数说明
- [ animSequence: List\<EasyTweenAnimationItem> ] 动画分镜序列
- [ duration: Duration ] 动画执行总时长
- [ curve: Curve ] 动画执行曲线，默认线性
- [ builder: Funtion ] 构建要使用动画的组件
- [ onStatus: Funtion ] 监听动画执行状态回调
- [ child: Widget ] 子组件
- [ delay: Duration ] 延时时间，默认0秒立即执行
- [ loop: bool ] 是否循环执行

#### EasyTweenAnimationItem 组件参数说明

- [ animatables: Map\<String, Animatable> ] 动画效果，该分镜时间段里的动画效果组
	- Map 说明
		- [ key : String tag ] 效果名，如 width
		- [ value : Animatable animatable ] 动画效果，一般使用 Tween 或者 ColorTween
- [ weight: double ] 分镜占动画执行总时长权重百分比(0-100)



### 加入群聊
扫码加我微信加入微信交流群（请备注：Flutter库easy_anim）

<img src="doc/assets/wechat_qr.jpg" width="200" >

#### 🤗 🤗 🤗 如果对您有帮助，请star

CHANGELOG.md
## [2.1.0] - Declarative description animation component release

* Publish to pub.dev

## [2.1.1] - fix bug dispose

* fix bug dispose

## [3.1.1] - Support null safety

* Support null safety


test/easy_anim_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:easy_anim/easy_anim.dart';


void main() {
  testWidgets('Simple animation test', (WidgetTester tester) async {

    EasyTweenAnimation anim = EasyTweenAnimation(
      animSequence: [
        EasyTweenAnimationItem(
            animatables: {
              "width": Tween<double>(begin: 100, end: 200),
            },
            weight: 100
        ),
      ],
      duration: Duration(seconds: 1),
      builder: (BuildContext context, CurvedAnimation curvedAnimation, Map<String, Animation> animationMap, AnimationController animationController, Widget? child){

        Animation width = animationMap['width']!;

        return AnimatedBuilder(animation: curvedAnimation, builder: (context, child){
          return Transform.rotate(
            angle: 0.5,
            child: Container(
              color: Colors.red,
              width: width.value,
              height: 300,
            ),
          );
        });
      },
    );

    expect(anim, isNotNull);

  });
}


lib/src/easy_tween_animation.dart


import 'package:flutter/material.dart';

typedef EasyTweenAnimationWidgetBuilder = Widget Function(BuildContext context, CurvedAnimation curvedAnimation, Map<String, Animation> animationMap, AnimationController animationController, Widget? child);
typedef EasyTweenAnimationOnStatus = Widget Function(AnimationStatus status, AnimationController controller);


/// @Desc  : 简单的方式使用Flutter Tween动画
/// 1，自维护controller，使用动画代码简洁
/// 2，支持延时执行动画
/// 3，支持循环执行动画
/// 4，支持交织动画（并行动画）
///
/// Created by 枫儿 on 2021/2/24.
/// @email：hsnndly@163.com
class EasyTweenAnimation extends StatefulWidget{


  final List<EasyTweenAnimationItem> animSequence;
  final Duration duration;
  final Curve curve;
  final EasyTweenAnimationWidgetBuilder builder;
  final EasyTweenAnimationOnStatus? onStatus;
  final Widget? child;
  final Duration delay;
  final bool loop;


  /// @Desc  : 创建一个动画组件，自维护controller，使用简单
  ///
  ///  ## 参数列表
  /// [animSequence] 动画分镜序列
  /// [duration] 动画执行总时长
  /// [curve] 动画执行曲线，默认线性
  /// [builder] 构建要使用动画的组件
  /// [onStatus] 监听动画执行状态回调
  /// [child] 子组件
  /// [delay] 延时时间，默认0秒立即执行
  /// [loop] 是否循环执行
  ///
  /// ## 使用示例
  ///
  /// ```dart
  /// EasyTweenAnimation(
  ///   animSequence: [
  ///     EasyTweenAnimationItem(
  ///       animatables: {
  ///         "angle": Tween<double>(begin: 0, end: 1 * pi),
  ///         "color": ColorTween(begin: Colors.blue, end: Colors.blue),
  ///         "width": Tween<double>(begin: 100, end: 100),
  ///       },
  ///       weight: 50.0, // 占总时长的50%，也就是2秒的0%~50%时的动画效果
  ///     ),
  ///
  ///
  ///     EasyTweenAnimationItem(
  ///       animatables: {
  ///         "color": ColorTween(begin: Colors.blue, end: Colors.red),
  ///         "width": Tween<double>(begin: 100, end: 200),
  ///       },
  ///       weight: 50.0, // 占总时长的50%，也就是2秒的50%~100%时的动画效果
  ///     ),
  ///   ],
  ///
  ///   duration: Duration(seconds: 2), // 执行动画总时长
  ///   delay: Duration(milliseconds: 200), // 延时多久执行，默认0秒立即执行
  ///   loop: true, // 是否循环执行动画
  ///   builder: (BuildContext context, CurvedAnimation curvedAnimation, Map<String, Animation> animationMap, AnimationController animationController, Widget child){
  ///
  ///     // 从animationMap里取出各个效果的Animation对象
  ///     Animation angle = animationMap["angle"];
  ///     Animation color = animationMap['color'];
  ///     Animation width = animationMap['width'];
  ///
  ///     // AnimatedBuilder 里使用 各个效果的Animation对象value即可
  ///     return AnimatedBuilder(animation: curvedAnimation, builder: (context, child){
  ///       return Transform.rotate(
  ///         angle: angle.value,
  ///         child: Container(
  ///           color: color.value,
  ///           width: width.value,
  ///           height: width.value,
  ///         ),
  ///       );
  ///     });
  ///   },
  /// )
  /// ```
  ///
  /// @author: 枫儿
  EasyTweenAnimation({
    Key? key,
    required this.animSequence,
    required this.duration,
    this.curve = Curves.linear,
    required this.builder,
    this.onStatus,
    this.child,
    this.delay = Duration.zero,
    this.loop = false,
  }): super(key: key);

  @override
  State<StatefulWidget> createState() => _EasyTweenAnimationState();
}


class _EasyTweenAnimationState extends State<EasyTweenAnimation> with SingleTickerProviderStateMixin{

  late AnimationController _animationController;
  Map<String, List<TweenSequenceItem>> _tagTweenSequenceItems = {};
  late CurvedAnimation _animation;
  Map<String, Animation> _animationMap = {};

  @override
  void initState() {
    super.initState();
    // 创建控制器
    _animationController = AnimationController(duration: this.widget.duration, vsync: this);
    _animation = CurvedAnimation(parent: _animationController, curve: this.widget.curve);

    // 添加监听
    _animation.addStatusListener((status) {
      // 回调状态
      if(this.widget.onStatus != null){
        this.widget.onStatus!(status, _animationController);
      }
    });

    this.makeUpAnimSequence(this.widget.animSequence); // 补齐 各个分镜 未存在的效果
    _tagTweenSequenceItems = this.buildTagTweenSequenceItems(this.widget.animSequence); // 创建Tag 的 TweenSequenceItem 数组

    // 创建 各个Tag效果的 TweenSequence
    _tagTweenSequenceItems.forEach((String tag, List<TweenSequenceItem> items) {
      Animation animation = TweenSequence(items).animate(_animationController);
      _animationMap[tag] = animation;
    });



    // 延时执行动画（默认0秒）
    new Future.delayed(this.widget.delay, (){

      if(this.widget.loop){
        _animationController.repeat();
      }else{
        _animationController.forward();
      }

    });
  }


  /// @Desc  : 生成所有Tag的TweenSequenceItem
  /// @author: 枫儿
  Map<String, List<TweenSequenceItem>> buildTagTweenSequenceItems(List<EasyTweenAnimationItem> animSequence){

    Map<String, List<TweenSequenceItem>> tagTweenSequenceItems = {};

    // 循环处理分镜
    for(int i=0; i<animSequence.length; i++){
      EasyTweenAnimationItem clipItem = animSequence[i];

      // 循环处理分镜内单个效果
      clipItem.animatables.forEach((String tag, Animatable animatable){
          if(tagTweenSequenceItems[tag] == null){ //如果不存在这个tag对应数组，创建空数组
            tagTweenSequenceItems[tag] = [];
          }

          TweenSequenceItem tweenSequenceItem = TweenSequenceItem(tween: animatable, weight: clipItem.weight);
          tagTweenSequenceItems[tag]!.add(tweenSequenceItem);
      });
    }

    return tagTweenSequenceItems;
  }


  /// @Desc  : 补齐[animSequence]各个分镜里缺失的动画效果，默认保持上一个
  /// @author: 枫儿
  makeUpAnimSequence(List<EasyTweenAnimationItem> animSequence){
    List<String> tags = this.getTags(animSequence);

    // 循环所有分镜
    for(int i=0; i<animSequence.length; i++){
      EasyTweenAnimationItem clipItem = animSequence[i];

      // 检查该分镜所有Tag是否存在
      for(int j=0; j<tags.length; j++){
        String tag = tags[j];

        // 不存在的Tag，查找前面分镜的相同Tag补上
        if(clipItem.existTag(tag) == false){
          clipItem.animatables[tag] = this.findLastTag(animSequence, i, tag);
        }
      }

    }
  }


  /// @Desc  : 寻找动画序列里某个节点前面最近一个存在该tag的节点，要是到第一个都没找到，就断言中断并警告
  /// @author: 枫儿
  Animatable findLastTag(List<EasyTweenAnimationItem> animSequence, int index, String tag){
    Animatable? animate;
    for(int i=index; i>=0; i--){
      EasyTweenAnimationItem item = animSequence[i];

      if(item.existTag(tag)){ //不为空则为找到了
        animate = item.animatables[tag]!;
        break;
      }
    }

    assert(
      animate != null,
      "EasyTweenAnimation 组件 animSequence 列表属性的第一个 EasyTweenAnimationItem 应该包含animSequence存在的所有效果作为默认值。 当前「${tag}」效果不存在，请在第一个EasyTweenAnimationItem内添加「${tag}」默认效果 \n\n"
      "The first EasyTweenAnimationItem in the animSequence list property of the EasyTweenAnimation component should contain all the effects that exist in the animSequence as default values. The current 「${tag}」 effect does not exist, please add the 「${tag}」 default effect in the first EasyTweenAnimationItem\n");

    return animate!;
  }



  /// @Desc  : 获取所有的Tag
  /// @author: 枫儿
  List<String> getTags(List<EasyTweenAnimationItem> animSequence){
    List<String> tags = [];

    for(int i=0; i<animSequence.length; i++){
      EasyTweenAnimationItem clipItem = animSequence[i];

      clipItem.animatables.keys.toList().forEach((itemTag) {
        if(tags.contains(itemTag) == false){
          tags.add(itemTag);
        }
      });
    }
    return tags;
  }



  @override
  Widget build(BuildContext context) {
    return this.widget.builder(context, _animation, _animationMap, _animationController, this.widget.child);
  }



  @override
  void dispose() {
    // 组件销毁，释放 controller
    _animationController.dispose();

    super.dispose();
  }

}

/// 动画分镜
/// Created by 枫儿 on 2021/2/24.
/// @email：hsnndly@163.com
class EasyTweenAnimationItem{

  Map<String, Animatable> animatables;
  double weight;

  /// @Desc  : 动画分镜
  /// [animatables] 动画效果，该分镜时间段里的动画效果组
  /// [weight] 分镜占动画执行总时长权重百分比(0-100)
  EasyTweenAnimationItem({
    Key? key,
    required this.animatables,
    required this.weight,
  });


  /// @Desc  : 判断分镜内是否存在某个Tag
  /// @author: 枫儿
  bool existTag(String tag){
    if(this.animatables[tag] != null){
      return true;
    }else{
      return false;
    }
  }


}


example/lib/main.dart
import 'dart:math';

import 'package:easy_anim/easy_anim.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'easy_anim example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);

  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;


  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title!),
      ),
      body: SingleChildScrollView(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [

            this.one(),
            this.two(),
            this.three(),
            this.four(),
            this.five(),

            SizedBox(height: 500,)
          ],
        ),
      ),
    );
  }


  /// @Desc  : 单个动画单次执行
  /// @author: 枫儿
  Widget one(){
    return Card(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          ListTile(
            leading: Icon(Icons.album),
            title: Text('单个动画'),
            subtitle: Text("红色方块单个动画，执行一次 \n 紫色方块延迟3秒执行一次"),
          ),

          /// example
          EasyTweenAnimation(
              animSequence: [
                EasyTweenAnimationItem(
                    animatables: {
                      "width": Tween(begin: 0.0, end: 200.0),
                    },
                    weight: 100,
                ),
              ],
              duration: Duration(seconds: 2),
              builder: (BuildContext context, CurvedAnimation curvedAnimation, Map<String, Animation> animationMap, AnimationController animationController, Widget? child){

                Animation width = animationMap['width']!;

                return AnimatedBuilder(animation: curvedAnimation, builder: (context, child){
                      return Container(
                        width: width.value,
                        height: 100,
                        decoration: BoxDecoration(
                          color: Colors.red
                        ),
                      );
                    },
                );
              }
          ),

          SizedBox(height: 20,),


          /// example delay
          EasyTweenAnimation(
              animSequence: [
                EasyTweenAnimationItem(
                  animatables: {
                    "width": Tween(begin: 0.0, end: 200.0),
                  },
                  weight: 100,
                ),
              ],
              duration: Duration(seconds: 2),
              delay: Duration(seconds: 3),
              builder: (BuildContext context, CurvedAnimation curvedAnimation, Map<String, Animation> animationMap, AnimationController animationController, Widget? child){

                Animation width = animationMap['width']!;

                return AnimatedBuilder(animation: curvedAnimation, builder: (context, child){
                  return Container(
                    width: width.value,
                    height: 100,
                    decoration: BoxDecoration(
                        color: Colors.deepPurpleAccent
                    ),
                  );
                },
                );
              }
          ),

          SizedBox(height: 20,)


        ],
      ),
    );
  }



  /// @Desc  : 单个动画循环执行
  /// @author: 枫儿
  Widget two(){
    return Card(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          ListTile(
            leading: Icon(Icons.album),
            title: Text('循环执行'),
            subtitle: Text("黄色方块单个动画，循环执行"),
          ),

          /// example
          EasyTweenAnimation(
              animSequence: [
                EasyTweenAnimationItem(
                  animatables: {
                    "width": Tween(begin: 0.0, end: 200.0),
                  },
                  weight: 100,
                ),
              ],
              duration: Duration(seconds: 2),
              loop: true, ///循环执行
              curve: Curves.bounceInOut,
              builder: (BuildContext context, CurvedAnimation curvedAnimation, Map<String, Animation> animationMap, AnimationController animationController, Widget? child){

                Animation width = animationMap['width']!;

                return AnimatedBuilder(animation: curvedAnimation, builder: (context, child){
                  return Container(
                    width: width.value,
                    height: 100,
                    decoration: BoxDecoration(
                        color: Colors.amber
                    ),
                  );
                },
                );
              }
          ),

          SizedBox(height: 20,)
        ],
      ),
    );
  }



  /// @Desc  : 串行执行多个动画
  /// @author: 枫儿
  Widget three(){
    return Card(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          ListTile(
            leading: Icon(Icons.album),
            title: Text('串行执行多个动画'),
            subtitle: Text("蓝色方块串行执行多个动画（宽度放大->颜色改变为红色）"),
          ),

          /// example
          EasyTweenAnimation(
              animSequence: [
                EasyTweenAnimationItem(
                  animatables: {
                    "width": Tween(begin: 0.0, end: 200.0),
                    "color": ColorTween(begin: Colors.blue, end: Colors.blue), /// 第一个 EasyTweenAnimationItem 应该包含animSequence存在的所有效果作为默认值
                  },
                  weight: 50, /// 占总时长的50%，也就是0%~50%时的动画效果
                ),

                EasyTweenAnimationItem(
                  animatables: {
                    "width": Tween(begin: 200.0, end: 200.0),
                    "color": ColorTween(begin: Colors.blue, end: Colors.red),
                  },
                  weight: 50, /// 占总时长的50%，也就是50%~100%时的动画效果
                ),
              ],
              duration: Duration(seconds: 2),
              loop: true, ///循环执行
              builder: (BuildContext context, CurvedAnimation curvedAnimation, Map<String, Animation> animationMap, AnimationController animationController, Widget? child){

                Animation width = animationMap['width']!;
                Animation color = animationMap['color']!;

                return AnimatedBuilder(animation: curvedAnimation, builder: (context, child){
                  return Container(
                    width: width.value,
                    height: 100,
                    decoration: BoxDecoration(
                        color: color.value,
                    ),
                  );
                },
                );
              }
          ),

          SizedBox(height: 20,)
        ],
      ),
    );
  }



  /// @Desc  : 并行执行多个动画（交织动画）
  /// @author: 枫儿
  Widget four(){
    return Card(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          ListTile(
            leading: Icon(Icons.album),
            title: Text('并行执行多个动画（交织动画）'),
            subtitle: Text("蓝色方块并行执行多个动画（宽度放大+颜色改变为红色）"),
          ),

          /// example
          EasyTweenAnimation(
              animSequence: [
                EasyTweenAnimationItem(
                  animatables: {
                    "width": Tween(begin: 0.0, end: 200.0),
                    "color": ColorTween(begin: Colors.blue, end: Colors.red),
                  },
                  weight: 100,
                ),

              ],
              duration: Duration(seconds: 2),
              loop: true, ///循环执行
              builder: (BuildContext context, CurvedAnimation curvedAnimation, Map<String, Animation> animationMap, AnimationController animationController, Widget? child){

                Animation width = animationMap['width']!;
                Animation color = animationMap['color']!;

                return AnimatedBuilder(animation: curvedAnimation, builder: (context, child){
                  return Container(
                    width: width.value,
                    height: 100,
                    decoration: BoxDecoration(
                        color: color.value,
                    ),
                  );
                },
                );
              }
          ),

          SizedBox(height: 20,)
        ],
      ),
    );
  }



  /// @Desc  : 串行+并行执行多个动画效果
  /// @author: 枫儿
  Widget five(){
    return Card(

      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          ListTile(
            leading: Icon(Icons.album),
            title: Text('串行+并行执行多个动画效果'),
            subtitle: Text("方块旋转缩放 然后 修改颜色缩放"),
          ),

          /// example
          EasyTweenAnimation(
              animSequence: [
                EasyTweenAnimationItem(
                  animatables: {
                    "angle": Tween<double>(begin: 0, end: 1 * pi),
                    "size": Tween(begin: 50.0, end: 200.0),
                    "color": ColorTween(begin: Colors.blue[50], end: Colors.blue[400]),
                    "radius": Tween<double>(begin: 20, end: 0),
                    "borderColor": ColorTween(begin: Colors.red, end: Colors.deepPurpleAccent),
                  },
                  weight: 50,
                ),

                EasyTweenAnimationItem(
                  animatables: {
                    "angle": Tween<double>(begin: 1 * pi, end: 1 * pi),
                    "size": Tween(begin: 200, end: 50.0),
                    "color": ColorTween(begin: Colors.blue[400], end: Colors.amber),
                    "radius": Tween<double>(begin: 0, end: 360),
                    "borderColor": ColorTween(begin: Colors.deepPurpleAccent, end: Colors.brown),
                  },
                  weight: 50,
                ),

              ],
              duration: Duration(seconds: 2),
              loop: true, ///循环执行
              builder: (BuildContext context, CurvedAnimation curvedAnimation, Map<String, Animation> animationMap, AnimationController animationController, Widget? child){

                Animation size = animationMap['size']!;
                Animation color = animationMap['color']!;
                Animation angle = animationMap['angle']!;
                Animation radius = animationMap['radius']!;
                Animation borderColor = animationMap['borderColor']!;


                return AnimatedBuilder(animation: curvedAnimation, builder: (context, child){
                  return Transform.rotate(
                      angle: angle.value,
                      child: Container(
                        width: size.value,
                        height: size.value,
                        decoration: BoxDecoration(
                          color: color.value,
                          borderRadius: BorderRadius.circular(radius.value),
                          border: Border.all(color: borderColor.value, width: 2),
                        ),
                      )
                  );

                },
                );
              }
          ),

          SizedBox(height: 20,)
        ],
      ),
    );
  }





}
